doc/src/challenges/0007-atomic-types.md

```diff
@@ -0,0 +1,83 @@
+# Challenge 7: Safety of Methods for Atomic Types and `ReentrantLock`
+
+- **Status:** Open
+- **Tracking Issue:** [#83](https://github.com/model-checking/verify-rust-std/issues/83)
+- **Start date:** *2024-09-09*
+- **End date:** *2024-12-10*
+
+-------------------
+
+## Goal
+
+Verify the safety of:
+- The methods for atomic types in `core::sync::atomic` 
+- The implementation of `ReentrantLock` in `std::sync::reentrant_lock`
+
+### Success Criteria
+
+
+#### Part 1: Atomic Types
+
+First, verify the safety of methods on atomic types.
+
+Write safety contracts for each of the `from_ptr` methods:
+
+- `AtomicBool::from_ptr`
+- `AtomicPtr::from_ptr`
+- `AtomicI8::from_ptr`
+- `AtomicU8::from_ptr`
+- `AtomicI16::from_ptr`
+- `AtomicU16::from_ptr`
+- `AtomicI32::from_ptr`
+- `AtomicU32::from_ptr`
+- `AtomicI64::from_ptr`
+- `AtomicU64::from_ptr`
+- `AtomicI128::from_ptr`
+- `AtomicU128::from_ptr`
+
+Specifically, encode the safety conditions marked `#Safety` in the comments above the methods.
+
+For the integer and boolean atomics, verify that these contracts hold on all bit-valid values. For `AtomicPtr`, verify that the contracts hold for pointees of size 0, 8, 16, 32, 64, 128, and at least one non-power-of-two.
+
+Then, write and verify safety contracts for the remaining unsafe functions:
+
+- `atomic_store`
```

Comment:
Which functions are these? Are these meant to be the methods in the atomic types? Or are these the intrinsics themselves?

Reply:
This atomics file is a bit funky because there's three layers of function calls here:

- The methods for the atomic types, which invoke helper functions at the bottom of the file whenever they want to perform atomic operations
- Helper functions that invoke atomic intrinsics
- The atomic intrinsics themselves

For example, `AtomicI8::fetch_add` and `AtomicI8::fetch_byte_add` are wrappers around the helper function `atomic_add`, which in turn is a wrapper around `intrinsics::atomic_xadd_relaxed`.

So the question is: which layer should have contracts? I chose the helper functions for this challenge, since the method calls are just wrappers around the helpers and Kani doesn't support contracts on intrinsics. 

I don't feel too strongly about this choice, though, especially since we're remaining tool-agnostic (and planning to add contracts support for intrinsics anyway). So I can change the challenge to add contracts for a different layer (or multiple layers) if we'd prefer.

Reply:
Upon reflection, I decided to change the contracts to be on the methods, since in the future the methods may not use the helpers, and we want to ensure that the contracts continue to hold even as the standard library changes.